proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,250 @@
+// Valid: f is skipped, Buf is bound to t0, RWBuf is bound to u0
+
+struct Eg2 {
+  float f;
+  Buffer<float> Buf;
+  RWBuffer<float> RWBuf;
+  RWBuffer<float> RWBuf2;
+  };
+Eg2 e2 : register(t0) : register(u0); 
+// Valid: f is skipped, Buf is bound to t0, RWBuf is bound to u0. RWBuf2 gets automatically assigned to u1 even though there is no explicit binding for u1.
+
+struct Eg3 {
+  float f;
+  Buffer<float> Buf;
+  }; 
+Eg3 e3 : register(t0) : register(u0);
+// Valid: Buf gets bound to t0. Buf will also be bound to u0.
+
+struct Eg4 {
+  struct Bar {
+    RWBuffer<int> a;
+    };
+    Bar b;
+};
+Eg4 e4 : register(t0) 
+// Valid: Bar, the struct within Eg4, has a valid resource that can be bound to t0. 
+
+struct Eg5 {
+  SamplerState s[3];
+};
+
+Eg5 e5 : register(s5);
+// Valid: the first sampler state object within Eg5's s is bound to slot 5
+
+struct Eg6 {
+  float f;
+}; 
+Eg6 e6 : register(t0) 
+// DefaultError warning: "No resource contained in struct 'Eg6' can be bound to register type 't'"
+
+struct Eg7 {
+  struct Bar {
+      float f;
+    }
+    Bar b;
+};
+Eg7 e7 : register(t0) 
+// DefaultError warning: "No resource contained in struct 'Eg7' can be bound to register type 't'"
+
+```
+
+Finally, if the candidate type is not a valid resource type and not a UDT, the
+final case will be entered. Types that are or contain a resource are known as
+"intangible". In this case, we are dealing with types that cannot be intangible.
+Types that can be immediately determined to not be intangible (that is, types that
+cannot be a resource type or cannot contain a resource type) are types like booleans,
+int, float, float4, etc. If the variable type is among any numerics or a type that
+cannot be an intangible type, there is only one exception where the register keyword
+can legally be applied to the variable. The only valid register type for such a 
+numeric variable type is 'c'. The register keyword in this statement:
+`float f : register(c0)`
+isn't binding a resource, rather it is specifying a constant register binding offset
+within the $Globals cbuffer, which is legacy behavior from DX9. Because this specific 
+register overload is only applicable if the developer wants to make modifications to the
+$Globals buffer, it is not useful when the overload is not used inside the $Globals
+context. That is, if this register overload appears within a `cbuffer {...}` or 
+`tbuffer {...}` scope, the register number will be ignored. In this case, a warning will
+be emitted that is treated as an error:
+`warning: variable of type '%0' bound to register type '%1' does not contain a matching '%select{SRV|UAV|CBV|Sampler}2' resource`
+All other register types applied to such a type will emit an error diagnostic that 
+`"'%0' is an invalid resource type for register type '%1'"`, except for 'b' or 'i'.
+If the register type is 'b' or 'i', a warning will be emitted instead, that will be 
+treated as an error by default. <br>
+Below are some examples:

```

Comment:
Not sure why embedded `<br>` is showing up here, but it's probably better to just use a newline instead.  There are other examples of this later as well.
```suggestion
treated as an error by default.

Below are some examples:
```